tools/inc/stljobs.h

```diff
@@ -0,0 +1,618 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#pragma once
+
+#include <algorithm>
+#include <assert.h>
+#include <atomic>
+#include <exception>
+#include <iterator>
+#include <limits.h>
+#include <memory>
+#include <random>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <string>
+#include <string_view>
+#include <utility>
+
+#include <Windows.h>
+#include <userenv.h>
+
+struct api_exception : std::exception {
+    const char* api;
+    unsigned long lastError;
+
+    explicit api_exception(const char* const api_, const unsigned long lastError_) noexcept
+        : api(api_), lastError(lastError_) {}
+
+    [[noreturn]] void give_up() const {
+        fflush(stdout);
+        fprintf(stderr, "The API \"%s\" failed unexpectedly; last error 0x%08lX\n", api, lastError);
+        abort();
+    }
+
+    [[nodiscard]] const char* what() const noexcept override {
+        return "win32 exception";
+    }
+};
+
+[[noreturn]] inline void api_failure(const char* const api, const unsigned long lastError = GetLastError()) {
+    throw api_exception{api, lastError};
+}
+
+void close_handle(const HANDLE toClose) noexcept {

```

Comment:
Can I chime in like a whole year later and say that a neat trick I discovered years ago is that by marking HANDLES like this parameter with `_Post_ptr_invalid_` you get diagnostics from code analysis if you reuse handles that are closed